packages/data/src/types.ts

```diff
@@ -170,9 +170,17 @@ export type ConfigOf< S > = S extends StoreDescriptor< infer C > ? C : never;
 
 export type ActionCreatorsOf< Config extends AnyConfig > =
 	Config extends ReduxStoreConfig< any, infer ActionCreators, any >
-		? ActionCreators
+		? PromisifiedActionCreator< ActionCreators >
 		: never;
 
+// When dispatching an action creator, the return value is a promise.
+type PromisifiedActionCreator< ActionCreators extends MapOf< ActionCreator > > =
+	{
+		[ Action in keyof ActionCreators ]: (
+			...args: Parameters< ActionCreators[ Action ] >
+		) => Promise< void >;
```

Comment:
Instead of `Promise< void >`, should this be `Promise< ReturnType< ActionCreators[ Action ] > >`? It's more precise type, thunk actions commonly return some value.

Reply:
Yeah, I was thinking about this! My concern was a use case like this:

```ts
// Action creator:
function doFoo( value: number ) {
  return { type: 'FOO', value };
}

const { doFoo } = useDispatch( store );

doFoo.then( value => {
  console.log( value ); // I assume undefined, but the type is actually { type: 'FOO', value: number } 
} );
```

Reply:
Thoughts on how to handle that?

Reply:
Actually, after reading the [redux docs](https://redux.js.org/api/store#dispatchaction), the return type is the actual action object.

Reply:
The convention is that plain Redux action dispatch indeed returns the action object itself, but for thunks the return value is the value returned from the thunk function. And generally, any Redux middleware that enhances the `dispatch` function can change the return type arbitrarily. In practice all middlewares only _add_ new possible return types to the existing one because they want to be compatible with each other.

> Can we tell if the action is a thunk?

In principle it should be possible. For the action creators in the `actions` field of the store config, we know what they return. Either a plain object or a thunk function. And we can infer types from that.

`redux-thunk` has types like `ThunkDispatch` and `ThunkAction` and uses them to correctly type dispatches:

https://github.com/reduxjs/redux-thunk/blob/master/src/types.ts#L14-L37